### ISSUE TITLE
Misleading distinction between 'return' and 'return ()' in EmptyReturn

### ISSUE THREAD

**Issue by vpit** [Grinnz/Perl-Critic-Freenode#20]
Hello,

This module advertises 'return ()' as correct and 'return' as incorrect, and actually reports the latter as a violation if I read the code correctly. This is misleading because these two forms are syntaxically completely equivalent. While I understand that this could be the object of an internal stylistic policy when **writing** the sub, it has no implication to the problem your Perl::Critic policy is trying to advert : being surprised by an unpredictable number of return values when **calling** the sub in list context.

I suggest you to collect all the return points of any sub and report a violation if they are possibly inconsistent (like having both a 'return' and a 'return $foo'). In particular, note that a conditional at the end of a sub yields a scalar value.

Vincent

**Comment by Grinnz** [comment]
You are correct that "return ()" is not any different, but there is no way to tell if "return" on its own is intended to be a list context return (where it would be correct) or a scalar context return (where it would be in violation). Similarly, there's no way to tell if "return $foo" is a scalar or list context return. I will consider your ideas however.

**Comment by Grinnz** [comment]
Some further things to mention: This issue has been bugging me for a while now, and I've considered just dropping the policy, if the vulnerability just did not come up so often. And some more things that must be considered:
1. I am OK with requiring "return()" in a sub which intends to return list context, as a self-documented explicit return. I know this is not a compelling point.
2. I would be OK with simply allowing instances of "return" when no "return (...)" is present, except this is complicated by trying to detect implicit returns, or when an expression or call to another subroutine is returned.
3. The real difficult situation is subs which do not intend to return anything, especially event loop callbacks, for instance. These will often have "return"s as breakout points and have no need for specifying any return value, but there is just no way to detect that.

**Comment by Grinnz** [comment]
Whoops, point 2 should read, when no "return $foo" is present

**Comment by Grinnz** [comment]
In an effort to at least avoid throwing violations in subroutines which are meant for void context, I updated the policy to report violations only when no non-empty return is found in the same subroutine. Unfortunately this also means it finds a lot less cases due to the complexity of figuring out what "the same subroutine" means, but it should hopefully report the basic cases. For the initially reported issue I mentioned in the documentation that "return" and "return ()" are functionally equivalent, but the latter is recommended for functions which intend to return list context.

It could be improved in the future to apply to anonymous subroutine scopes (only applies to named subroutine definitions), and to find "return" statements within more types of blocks (skips over any block which is not a compound statement like "if" or "foreach"). Either of these would involve somewhat complex logic but should be doable.
